File: ios/Oriveo/Oriveo/Core/Providers/ModelDisplayLookup.swift
import Foundation

nonisolated struct ModelDisplayLookup: Sendable {
    nonisolated fileprivate struct Entry: Sendable {
        let providerKind: ProviderKind
        let providerDisplayName: String
        let modelDisplayNamesByKey: [String: String]
        let contextLengthsByKey: [String: Int]
        let relayKind: RelayKind?
        /// The enabled models this entry was built from. When a conversation row falls back to searching the
        /// enabled models, the index below is used only if the caller's Provider holds this same array (shared
        /// storage makes the comparison O(1)); otherwise it compares one by one as before.
        let enabledModels: [AIModel]
        let enabledModelIndex: SharedMatchingModelIndex
    }

    /// A shared `ModelResolver.MatchingModelIndex`: the lookup is a value type passed between views, so the index is
    /// built once on demand and then shared by every holder. It is locked because a background prewarm and a
    /// main-thread query can overlap.
    nonisolated final class SharedMatchingModelIndex: @unchecked Sendable {
        private let models: [AIModel]
        private let providerKind: ProviderKind
        private let lock = NSLock()
        private var index: ModelResolver.MatchingModelIndex?

        init(_ models: [AIModel], providerKind: ProviderKind) {
            self.models = models
            self.providerKind = providerKind
        }

        func match(_ modelID: String) -> AIModel? {
            lock.lock()
            defer { lock.unlock() }
            var built = index ?? ModelResolver.MatchingModelIndex(models, providerKind: providerKind)
            let result = built.match(modelID)
            index = built
            return result
        }

        /// Builds both the exact and the fuzzy layer now.
        func prewarm() {
            lock.lock()
            defer { lock.unlock() }
            var built = index ?? ModelResolver.MatchingModelIndex(models, providerKind: providerKind)
            built.prepareStoredIdentifierIndex()
            index = built
        }
    }

    private let entriesByProviderID: [UUID: Entry]
    private let metadata: MetadataClient

    private init(entriesByProviderID: [UUID: Entry], metadata: MetadataClient) {
        self.entriesByProviderID = entriesByProviderID
        self.metadata = metadata
    }

    nonisolated init(providers: [Provider], metadata: MetadataClient = .shared) {
        self.metadata = metadata
        self.entriesByProviderID = ModelDisplayLookupBuilder.makeEntries(
            providers: providers,
            metadata: metadata
        )
    }

    nonisolated static let empty = ModelDisplayLookup(entriesByProviderID: [:], metadata: .shared)

    nonisolated static func fingerprint(
        providers: [Provider],
        metadata: MetadataClient = .shared
    ) -> Int {
        var hasher = Hasher()
        hasher.combine(metadata.syncMetadataSnapshotVersion() ?? -1)
        hasher.combine(providers.count)
        for provider in providers {
            hasher.combine(provider.id)
            hasher.combine(provider.kind)
            hasher.combine(provider.customName)
            hasher.combine(provider.kind == .relay ? provider.relayKind : nil)
            hasher.combine(provider.models.count)
            for model in provider.models {
                hasher.combine(model.id)
                hasher.combine(model.name)
                hasher.combine(model.canonicalModelId)
                hasher.combine(model.contextLength)
            }
            for model in provider.catalogModels {
                hasher.combine(model.id)
                hasher.combine(model.name)
                hasher.combine(model.canonicalModelId)
                hasher.combine(model.contextLength)
            }
        }
        return hasher.finalize()
    }

    /// Exactly equivalent to `ModelResolver.matchingModel(modelID:in: provider.models, providerKind:)`. When this
    /// lookup was built from `provider`'s current enabled models it queries the index (after "Add all" the enabled
    /// models are the whole relay catalog, and comparing one by one per row grows with the catalog); otherwise it
    /// compares one by one as before.
    nonisolated func matchingEnabledModel(modelID: String, in provider: Provider) -> AIModel? {
        guard let entry = entriesByProviderID[provider.id],
              entry.providerKind == provider.kind,
              entry.enabledModels == provider.models else {
            return ModelResolver.matchingModel(modelID: modelID, in: provider.models, providerKind: provider.kind)
        }
        return entry.enabledModelIndex.match(modelID)
    }

    /// Builds every Provider's enabled-model index now. Call it off the main thread.
    nonisolated func prewarm() {
        for entry in entriesByProviderID.values {
            entry.enabledModelIndex.prewarm()
        }
    }

    nonisolated func providerDisplayName(providerID: UUID) -> String? {
        entriesByProviderID[providerID]?.providerDisplayName
    }

    nonisolated func relayKind(providerID: UUID) -> RelayKind? {
        entriesByProviderID[providerID]?.relayKind
    }

    nonisolated func modelDisplayName(
        providerID: UUID,
        modelID: String,
        fallback: String? = nil
    ) -> String? {
        guard let entry = entriesByProviderID[providerID] else { return fallback }
        let trimmedModelID = modelID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedModelID.isEmpty else { return fallback }

        for candidate in ModelResolver.normalizedLookupCandidates(
            for: trimmedModelID,
            providerKind: entry.providerKind
        ) {
            if let displayName = entry.modelDisplayNamesByKey[candidate.lowercased()] {
                return displayName
            }
        }

        if entry.providerKind != .relay,
           let resolved = metadata.syncResolveCatalogModel(
            modelID: trimmedModelID,
            providerKind: entry.providerKind
           ) {
            return resolved.displayName ?? fallback
        }

        return fallback
    }

    nonisolated func contextLength(providerID: UUID, modelID: String) -> Int? {
        guard let entry = entriesByProviderID[providerID] else { return nil }
        let trimmedModelID = modelID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedModelID.isEmpty else { return nil }

        for candidate in ModelResolver.normalizedLookupCandidates(
            for: trimmedModelID,
            providerKind: entry.providerKind
        ) {
            if let contextLength = entry.contextLengthsByKey[candidate.lowercased()] {
                return contextLength
            }
        }

        if entry.providerKind != .relay,
           let resolved = metadata.syncResolveCatalogModel(
            modelID: trimmedModelID,
            providerKind: entry.providerKind
           ) {
            return resolved.contextLength
        }

        return nil
    }
    fileprivate nonisolated static func localLookupKeys(for model: AIModel, providerKind: ProviderKind) -> [String] {
        let rawValues = [
            model.id,
            ModelResolver.resolvedProviderModelIdentifier(model.id, providerKind: providerKind),
            model.canonicalModelId,
            model.canonicalModelId.map {
                ModelResolver.resolvedProviderModelIdentifier($0, providerKind: providerKind)
            }
        ].compactMap { $0 }

        var keys: [String] = []
        for rawValue in rawValues {
            for candidate in ModelResolver.normalizedLookupCandidates(
                for: rawValue,
                providerKind: providerKind
            ) {
                let normalized = candidate.lowercased()
                guard !normalized.isEmpty else { continue }
                guard !keys.contains(normalized) else { continue }
                keys.append(normalized)
            }
        }
        return keys
    }
}

private enum ModelDisplayLookupBuilder {
    nonisolated static func makeEntries(
        providers: [Provider],
        metadata: MetadataClient
    ) -> [UUID: ModelDisplayLookup.Entry] {
        Dictionary(uniqueKeysWithValues: providers.map { provider in
            (
                provider.id,
                ModelDisplayLookup.Entry(
                    providerKind: provider.kind,
                    providerDisplayName: resolvedProviderDisplayName(
                        provider: provider,
                        metadata: metadata
                    ),
                    modelDisplayNamesByKey: makeModelDisplayNames(
                        provider: provider,
                        metadata: metadata
                    ),
                    contextLengthsByKey: makeContextLengths(
                        provider: provider,
                        metadata: metadata
                    ),
                    relayKind: provider.kind == .relay ? provider.relayKind : nil,
                    enabledModels: provider.models,
                    enabledModelIndex: ModelDisplayLookup.SharedMatchingModelIndex(
                        provider.models,
                        providerKind: provider.kind
                    )
                )
            )
        })
    }

    private nonisolated static func resolvedProviderDisplayName(
        provider: Provider,
        metadata: MetadataClient
    ) -> String {
        if provider.kind == .relay {
            return provider.customName ?? ProviderKind.relay.displayName
        }
        return metadata.syncProviderDisplayName(providerKind: provider.kind) ?? provider.kind.displayName
    }

    private nonisolated static func sourceModels(for provider: Provider) -> [AIModel] {
        provider.kind == .relay
            ? (provider.catalogModels.isEmpty ? provider.models : provider.catalogModels + provider.models)
            : provider.models + provider.catalogModels
    }

    private nonisolated static func makeContextLengths(
        provider: Provider,
        metadata: MetadataClient
    ) -> [String: Int] {
        var contextLengthsByKey: [String: Int] = [:]
        for model in sourceModels(for: provider) {
            let resolved = metadata.syncResolveCatalogModel(
                modelID: model.id,
                providerKind: provider.kind
            )?.contextLength ?? model.contextLength
            guard let resolved, resolved > 0 else { continue }

            for key in ModelDisplayLookup.localLookupKeys(for: model, providerKind: provider.kind) {
                contextLengthsByKey[key] = resolved
            }
        }
        return contextLengthsByKey
    }

    private nonisolated static func makeModelDisplayNames(
        provider: Provider,
        metadata: MetadataClient
    ) -> [String: String] {
        let sourceModels = sourceModels(for: provider)

        var namesByKey: [String: String] = [:]
        for model in sourceModels {
            let resolvedDisplayName = metadata.syncResolveCatalogModel(
                modelID: model.id,
                providerKind: provider.kind
            )?.displayName ?? model.name

            for key in ModelDisplayLookup.localLookupKeys(for: model, providerKind: provider.kind) {
                namesByKey[key] = resolvedDisplayName
            }
        }
        return namesByKey
    }
}

File: android/app/src/main/java/ai/oriveo/community/core/provider/ModelDisplayLookup.kt
package ai.oriveo.community.core.provider

import androidx.compose.runtime.Stable
import ai.oriveo.community.core.data.remote.MetadataClient
import ai.oriveo.community.core.model.AIModel
import ai.oriveo.community.core.model.Provider
import ai.oriveo.community.core.model.ProviderKind
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.flow.mapLatest

/**
 * Lightweight read path for display names: it resolves a provider name, a model
 * name and the canonical model id, and nothing else.
 *
 * Deliberately narrow: it never returns a full catalog, never triggers
 * [ProviderCatalogResolver], and knows nothing about recommendations, grouping or
 * enabled state.
 *
 * Every conversation row on Home and every message in a chat asks it once, and a relay
 * catalog comes from the user's own server (a public catalog can hold 22,000 models). So
 * each provider's local model index is built once per instance and every row after that is
 * a lookup instead of a scan. Screens get instances from [modelDisplayLookups], which builds
 * one per providers emission in the background and [prewarm]s it, so composition only reads a
 * finished one; the indexes go away with the instance and never answer from an old catalog.
 *
 * The indexes are internal memoization that changes no observable result, so the class is
 * still stable for Compose.
 */
@Stable
class ModelDisplayLookup private constructor(
    providers: List<Provider>,
    private val metadata: MetadataClient,
    /** False only for [Pending]: the screen has no lookup built from any emission yet, and callers must not show its fallback as a display name. */
    val isReady: Boolean,
) {
    constructor(
        providers: List<Provider>,
        metadata: MetadataClient = MetadataClient.instance,
    ) : this(providers, metadata, isReady = true)

    private val providersById = providers.associateBy { it.id }
    private val localIndexes = HashMap<String, ModelSelectionUtils.CatalogMatchIndex>()
    private val enabledIndexes = HashMap<String, ModelSelectionUtils.CatalogMatchIndex>()

    fun providerDisplayName(providerId: String?): String? =
        providerId?.let { providersById[it]?.displayName }

    fun modelDisplayName(
        providerId: String?,
        modelId: String?,
        fallback: String?,
    ): String? {
        val provider = providerId?.let { providersById[it] } ?: return fallback
        val targetId = modelId?.trim().orEmpty()
        if (targetId.isEmpty()) return fallback

        if (provider.kind == ProviderKind.Relay) {
            localModel(provider, targetId)?.name?.let { return it }
        } else {
            val metadataResolved = metadata.resolveCatalogModel(targetId, provider.kind)
            metadataResolved?.displayName?.takeIf { it.isNotBlank() }?.let { return it }
            metadataResolved?.canonicalModelId?.let { canonicalId ->
                localModel(provider, canonicalId)?.name?.let { return it }
                metadata.resolveCatalogModel(canonicalId, provider.kind)
                    ?.displayName
                    ?.takeIf { it.isNotBlank() }
                    ?.let { return it }
            }
            localModel(provider, targetId)?.name?.let { return it }
        }

        return fallback
    }

    fun canonicalModelId(
        providerId: String?,
        modelId: String?,
    ): String? {
        val provider = providerId?.let { providersById[it] } ?: return null
        val targetId = modelId?.trim().orEmpty()
        if (targetId.isEmpty()) return null

        localModel(provider, targetId)?.canonicalModelId?.takeIf { it.isNotBlank() }?.let { return it }
        localModel(provider, targetId)?.id?.let { return it }

        if (provider.kind != ProviderKind.Relay) {
            metadata.resolveCatalogModel(targetId, provider.kind)?.canonicalModelId?.let { return it }
        }

        return null
    }

    /**
     * Returns exactly what `ProviderSelectionSnapshot.selectedModel(provider, modelId, metadata)` returns, looking
     * enabled models up in this instance's index. After "add all" the enabled models are the whole relay catalog,
     * and comparing them one by one for every row grows with the catalog. When [provider] is not the object this
     * instance indexed (the caller holds another emission), it falls back to the one-by-one comparison.
     */
    fun selectedModel(
        provider: Provider,
        modelId: String?,
        metadata: MetadataClient = MetadataClient.instance,
    ): AIModel? {
        if (providersById[provider.id] !== provider) {
            return ProviderSelectionSnapshot.selectedModel(provider, modelId, metadata)
        }
        return ProviderSelectionSnapshot.selectedModel(provider, modelId, metadata) { _, targetId ->
            enabledIndex(provider).match(targetId)
        }
    }

    /** Builds every provider's local and enabled indexes (fuzzy layer included) now. Call it off the main thread; queries after that are table lookups. */
    fun prewarm() {
        providersById.values.forEach { provider ->
            localIndex(provider).prewarm()
            enabledIndex(provider).prewarm()
        }
    }

    /** Returns exactly what `matchingModel(localCandidates, targetId)` returns, with the same candidates in the same order. */
    private fun localModel(provider: Provider, targetId: String): AIModel? =
        localIndex(provider).match(targetId)

    private fun localIndex(provider: Provider): ModelSelectionUtils.CatalogMatchIndex = synchronized(localIndexes) {
        localIndexes.getOrPut(provider.id) {
            val localCandidates = if (provider.catalogModels.isNotEmpty()) {
                (provider.models + provider.catalogModels).distinctBy { it.id }
            } else {
                provider.models
            }
            ModelSelectionUtils.catalogMatchIndex(localCandidates)
        }
    }

    private fun enabledIndex(provider: Provider): ModelSelectionUtils.CatalogMatchIndex = synchronized(enabledIndexes) {
        enabledIndexes.getOrPut(provider.id) {
            ModelSelectionUtils.catalogMatchIndex(ProviderSelectionSnapshot.enabledModels(provider))
        }
    }

    companion object {
        /** Placeholder until a lookup has been built from any emission. */
        val Pending: ModelDisplayLookup by lazy { ModelDisplayLookup(emptyList(), MetadataClient.instance, isReady = false) }
    }
}

/**
 * Builds a [ModelDisplayLookup] for each providers emission on [dispatcher] and calls [ModelDisplayLookup.prewarm], so
 * Home, folders and chat only read a finished lookup during composition. Indexing a 22k relay catalog takes tens of
 * milliseconds of CPU, which the first row used to pay on the main thread after `remember(providers)`. A new emission
 * discards an unfinished older build.
 */
@OptIn(ExperimentalCoroutinesApi::class)
fun Flow<List<Provider>>.modelDisplayLookups(
    dispatcher: CoroutineDispatcher = Dispatchers.Default,
): Flow<ModelDisplayLookup> = mapLatest { providers ->
    ModelDisplayLookup(providers).also { it.prewarm() }
}.flowOn(dispatcher)
